### PR TITLE
ASC-402: Modify NSTemplateTierSpec to support SA token copy

### PR DESCRIPTION
**Pull request by Kartikey-star**
## Description
A few sentences describing the overall goals of the pull request's commits.

## Checks
1. Did you run `make generate` target? Yes

2. Did `make generate` change anything in other projects (host-operator, member-operator)? Yes

3. In case of **new** CRD, did you the following? **yes/no**
    - make/generate.mk in this repository
    - `resources/setup/roles/host.yaml` in the sandbox-sre repository
    - `PROJECT` file: https://github.com/codeready-toolchain/host-operator/blob/master/PROJECT
    - `CSV` file: https://github.com/codeready-toolchain/host-operator/blob/master/config/manifests/bases/host-operator.clusterserviceversion.yaml

4. In case other projects are changed, please provides PR links.
    - host-operator: https://github.com/codeready-toolchain/host-operator/pull/886

### CODE REVIEW

**Review by MatousJobanek — COMMENTED** [review]
I don't see any of the generated files to be modified as part of this PR - when you run `make generate` then it should update `zz_generated.deepcopy.go` `zz_generated.openapi.go` (doesn't have to be both - it depends on the type of the changes)

File: api/v1alpha1/nstemplatetier_types.go [hunk 1]
```diff
@@ -22,6 +22,10 @@ type NSTemplateTierSpec struct {
 	// +optional
 	// +mapType=atomic
 	SpaceRoles map[string]NSTemplateTierSpaceRole `json:"spaceRoles,omitempty"`
+
+	// flag to signify whether to copy SA token or not
+	// +optional
+	CopySaToken bool `json:"copySaToken,omitempty"`
```
[the remark this answers is not in the document]
Comment (Kartikey-star) [reply api/v1alpha1/nstemplatetier_types.go:28]: Let's go with the last one . Looks better.
type SpaceRequestConfig struct {
	// +optional
	SaTokenToCopy string `json:"saTokenToCopy,omitempty"`
}

**Comment by Kartikey-star** [comment]
> I don't see any of the generated files to be modified as part of this PR - when you run `make generate` then it should update `zz_generated.deepcopy.go` `zz_generated.openapi.go` (doesn't have to be both - it depends on the type of the changes)
 I have named my fork as https://github.com/Kartikey-star/toolchain-api...Due to this i believe the changes in `zz_generated.openapi.go` were showing in a different directory which does not exist on master

**Comment by mfrancisc** [comment]
> I have named my fork as https://github.com/Kartikey-star/toolchain-api...Due to this i believe the changes in zz_generated.openapi.go were showing in a different directory which does not exist on master

@Kartikey-star yes, it's recommended to clone all the projects into this folder structure in order for the `make generate` target to work:
```
~/go/src/github.com/codeready-toolchain/api
```
same for the other repos host-operator|registration-service|member-operator|toolchain-e2e ....

**Comment by Kartikey-star** [comment]
Thanks @mfrancisc , I had cloned the repos outside my GOPATH..Fixing that seems to work.

File: api/v1alpha1/nstemplatetier_types.go [hunk 2]
```diff
@@ -22,6 +22,15 @@ type NSTemplateTierSpec struct {
 	// +optional
 	// +mapType=atomic
 	SpaceRoles map[string]NSTemplateTierSpaceRole `json:"spaceRoles,omitempty"`
+
+	// Provides the name of the Service Account whose token is to be copied
+	// +optional
+	SpaceRequestConfigName SpaceRequestConfig `json:"spaceRequestConfig,omitempty"`
+}
+
+type SpaceRequestConfig struct {
```
Comment (MatousJobanek) [inline api/v1alpha1/nstemplatetier_types.go:31]:
```suggestion
// +k8s:openapi-gen=true
type SpaceRequestConfig struct {
```

**Review by MatousJobanek — COMMENTED** [review]
could you please add/fix the comments of the params and the new struct to describe the purpose of it?

File: api/v1alpha1/nstemplatetier_types.go [hunk 3]
```diff
@@ -22,6 +22,17 @@ type NSTemplateTierSpec struct {
 	// +optional
 	// +mapType=atomic
 	SpaceRoles map[string]NSTemplateTierSpaceRole `json:"spaceRoles,omitempty"`
+
+	// Provides the name of the Service Account whose token is to be copied
+	// +optional
+	SpaceRequestConfigName SpaceRequestConfig `json:"spaceRequestConfig,omitempty"`
```
Comment (mfrancisc) [inline api/v1alpha1/nstemplatetier_types.go:28]:
```suggestion
	SpaceRequestConfigName *SpaceRequestConfig `json:"spaceRequestConfig,omitempty"`
```
for `omitempty` to work it has to be a pointer. I faced a similar issue here https://github.com/codeready-toolchain/api/pull/382

Comment (Kartikey-star) [reply api/v1alpha1/nstemplatetier_types.go]: Yeah, thanks. Updated.

File: api/v1alpha1/nstemplatetier_types.go [hunk 4]
```diff
@@ -22,6 +22,17 @@ type NSTemplateTierSpec struct {
 	// +optional
 	// +mapType=atomic
 	SpaceRoles map[string]NSTemplateTierSpaceRole `json:"spaceRoles,omitempty"`
+
+	// Provides the name of the Service Account whose token is to be copied
```
Comment (mfrancisc) [inline api/v1alpha1/nstemplatetier_types.go:26]: minor thing - but I guess this comment relates more to the `SaTokenToCopy` field, while the struct comment could be something generic like: "SpaceRequestConfigName stores all the configuration related to the Space Request feature". 

WDYT?

Comment (mfrancisc) [reply api/v1alpha1/nstemplatetier_types.go:26]: same for the comment on the struct at line 31.

Comment (Kartikey-star) [reply api/v1alpha1/nstemplatetier_types.go]: That seems better to me .Updated .

**Review by mfrancisc — APPROVED** [review]
Looks good 👍 

Thanks for addressing my comments!

File: api/v1alpha1/nstemplatetier_types.go [hunk 5]
```diff
@@ -22,6 +22,18 @@ type NSTemplateTierSpec struct {
...
+	// +optional
+	SpaceRequestConfigName *SpaceRequestConfig `json:"spaceRequestConfig,omitempty"`
+}
+
+// SpaceRequestConfig contains all the configuration related to the Space Request feature
+// +k8s:openapi-gen=true
+type SpaceRequestConfig struct {
+	// Provides the name of the Service Account whose token is to be copied
+	// +optional
+	SaTokenToCopy string `json:"saTokenToCopy,omitempty"`
```
Comment (xcoulon) [inline api/v1alpha1/nstemplatetier_types.go:36]:
```suggestion
	ServiceAccountName string `json:"serviceAccountName,omitempty"`
```
`SaTokenToCopy` _may_ suggest that the field contains the actual token

Comment (Kartikey-star) [reply api/v1alpha1/nstemplatetier_types.go]: Yeah , updated

File: api/v1alpha1/nstemplatetier_types.go [hunk 6]
```diff
@@ -22,6 +22,18 @@ type NSTemplateTierSpec struct {
 	// +optional
 	// +mapType=atomic
 	SpaceRoles map[string]NSTemplateTierSpaceRole `json:"spaceRoles,omitempty"`
+
+	// SpaceRequestConfigName stores all the configuration related to the Space Request feature
+	// +optional
+	SpaceRequestConfigName *SpaceRequestConfig `json:"spaceRequestConfig,omitempty"`
```
Comment (MatousJobanek) [inline api/v1alpha1/nstemplatetier_types.go:28]: Why is there the "Name" suffix?
It should be
```suggestion
	// SpaceRequestConfig stores all the configuration related to the Space Request feature
	// +optional
	SpaceRequestConfig *SpaceRequestConfig `json:"spaceRequestConfig,omitempty"`
```

**Review by MatousJobanek — COMMENTED** [review]
I just noticed that there is `.DS_Store` included in the PR - what is it? could you please remove it?

**Comment by Kartikey-star** [comment]
> I just noticed that there is `.DS_Store` included in the PR - what is it? could you please remove it?

This file does exists on the main branch itself https://github.com/codeready-toolchain/api, it got altered hence committed. Shall i delete it?

**Review by MatousJobanek — APPROVED** [review]
thanks a lot for the additional changes :pray: :1st_place_medal:

**Pull request merged**